### PR TITLE
Simple Rust driver that touches real hardware: PR 4/6

### PR DESCRIPTION
**Pull request by TheSven73**
#### Simple Rust driver that touches real h/w: step 4 of 6

@alex and @ojeda suggested that #254 should be split into smaller PRs, so they can be reviewed and merged individually.

Roadmap:

- [x] platform_device
- [x] devicetree (of) matching
- [x] probe
- [ ] **DrvData (you are here)**
- [ ] regmap/iomem
- [ ] finally the actual hwrng driver

### CODE REVIEW

**Comment by ksquirrel** [comment]
Review of `fa1e44af1aca`:
  - ✔️ Commit [`2727379`](https://github.com/Rust-for-Linux/linux/commit/27273792c9972b18fb3e84bab80e5f9460d7ecb8): Looks fine!
  - ✔️ Commit [`fa1e44a`](https://github.com/Rust-for-Linux/linux/commit/fa1e44af1aca236d5f251a28d1195f3fff0b1471): Looks fine!

File: drivers/char/hw_random/bcm2835_rng_rust.rs [hunk 1]
```diff
@@ -20,15 +26,41 @@ module! {
...
+impl FileOperations for RngDevice {
+    kernel::declare_file_operations!(read);
+
+    fn read<T: IoBufferWriter>(&self, _: &File, data: &mut T, offset: u64) -> Result<usize> {
+        // Succeed if the caller doesn't provide a buffer or if not at the start.
+        if data.is_empty() || offset != 0 {
+            return Ok(0);
+        }
+
+        data.write(&0_u32)?;
```
Comment (bjorn3) [inline drivers/char/hw_random/bcm2835_rng_rust.rs:46]: Maybe add a FIXME just in case?

Comment (TheSven73) [reply drivers/char/hw_random/bcm2835_rng_rust.rs:46]: Thanks for the review! While the PR x/6 effort is ongoing, the driver is a continual WIP. I'm just adding functionality, as it becomes available in the Rust kernel core.

Do you think there is merit in placing a "warning this is a work in progress, not fully functional" comment on top of the driver?

Comment (bjorn3) [reply drivers/char/hw_random/bcm2835_rng_rust.rs:46]: Maybe?

Comment (TheSven73) [reply drivers/char/hw_random/bcm2835_rng_rust.rs:46]: Sounds reasonable to me. I'll add it.

**Comment by ksquirrel** [comment]
Review of `d0472669f89e`:
  - ✔️ Commit [`669248b`](https://github.com/Rust-for-Linux/linux/commit/669248bdabc6325bd0256ef03126f4f8094910cf): Looks fine!
  - ✔️ Commit [`d047266`](https://github.com/Rust-for-Linux/linux/commit/d0472669f89e7db443759b64dc4773112c9f59da): Looks fine!

**Comment by TheSven73** [comment]
### v1 -> v2:
- rebased against latest `rust` branch

**Comment by TheSven73** [comment]
### v2 -> v3:
- rebased against latest `rust` branch

**Comment by TheSven73** [comment]
I don't think this PR is particularly controversial or complex? It'd be great if more people could review, so the whole PR series could progress.

File: rust/kernel/platdev.rs [hunk 2]
```diff
@@ -30,13 +30,31 @@ pub struct Registration {
 // (it is fine for multiple threads to have a shared reference to it).
 unsafe impl Sync for Registration {}
 
+extern "C" {
+    #[allow(improper_ctypes)]
```
Comment (alex) [inline rust/kernel/platdev.rs:34]: What's the improper ctype here?

**Review by alex — COMMENTED** [review]
One small question, otherwise LGTM.

File: rust/kernel/platdev.rs (hunk 2 above)
Comment (TheSven73) [reply rust/kernel/platdev.rs:34]: Without it, I get the following warning:
```rust
  RUSTC L rust/kernel.o
warning: `extern` block uses type `arch_spinlock_t`, which is not FFI-safe
    --> rust/kernel/platdev.rs:36:15
     |
36   |         pdev: *const bindings::platform_device,
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not FFI-safe
     |
     = note: `#[warn(improper_ctypes)]` on by default
     = help: consider adding a member to this struct
     = note: this struct has no fields
note: the type is defined here
    --> /home/sva/repos/kernel-arcx/rust/bindings_generated.rs:7331:1
     |
7331 | pub struct arch_spinlock_t {}
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: 1 warning emitted
```

Comment (nbdd0121) [reply rust/kernel/platdev.rs:34]: What type is using `arch_spinlock_t`? Is it `raw_spinlock`? You can add the type to the opaque list: https://github.com/Rust-for-Linux/linux/blob/d8c6b9ce7fb78375460422ba15d763c22f70d0af/rust/bindgen_parameters#L10.

Comment (TheSven73) [reply rust/kernel/platdev.rs:34]: Hi Gary, thanks for the help! I suspect that on arm 32-bit, `arch_spinlock_t` is just its own type, and not a typedef. So should I add it to `--opaque-type`?
https://github.com/Rust-for-Linux/linux/blob/d8c6b9ce7fb78375460422ba15d763c22f70d0af/arch/arm/include/asm/spinlock_types.h#L11-L24

Comment (TheSven73) [reply rust/kernel/platdev.rs:34]: Ok, when I add `arch_spinlock_t` to `--opaque-type`, I seem to enter some kind of rabbit-hole where different structures trigger the warning, and when you add those to `--opaque-type`, there are more structures that trigger it. So far I have:
```
--opaque-type arch_spinlock_t
--opaque-type lock_class_key
--opaque-type arch_rwlock_t
--opaque-type vm_userfaultfd_ctx
```
and still the warnings keep coming.

Comment (nbdd0121) [reply rust/kernel/platdev.rs:34]: Setting `arch_spinlock_t` itself to opaque type just propagate the issue one struct up :)

Do you need to peek into fields of `platform_device`? If not, just setting that to opaque type will be sufficient.

**Review by alex — APPROVED** [review]

File: rust/kernel/platdev.rs (hunk 2 above)
Comment (TheSven73) [reply rust/kernel/platdev.rs:34]: Edited wait a minute... yes, we do need to peek into `platform_device`. So that won't solve the issue...

Comment (TheSven73) [reply rust/kernel/platdev.rs:34]: But the peeking is just a hack, and will disappear once we proceed. I'll make a mental note to add `platform_device` to `--opaque-type` later, and remove the `#[allow(improper_ctypes)]`.

**Comment by TheSven73** [comment]
Thank you @alex, appreciate your kind assistance :)

**Pull request merged**